### ISSUE TITLE
Saveable Games

### ISSUE THREAD

**Issue by towerofnix** [ScratchAddons/ScratchAddons#1990]
**Describe the new addon**

As discussed in the previous (and much tamer) suggestion #1988, lists (particularly list monitors) are the de facto way to transfer data between Scratch and external programs. This is particularly valuable for saving progress in games; being able to export data and re-import it later is key for especially long games, when a plain password is too simple to suffice.

So, I think an addon which makes the user interface around lists much simpler and more versatile would be a neat addition! My thinking is, when a list monitor is shown (and perhaps only if its name contains the word "save", for example), an extra button would be added in the top right corner for saving and loading. This would open a (hopefully minimal) UI; you'd have the option to create a new save, and a list of previous saves (which you could also overwrite if you like). Each "save" is really just an object storing the data in a list; it would also have an optional property for setting a name (i.e. UI-visible comment), and the date it was created (and/or modified) at too. If you choose a save to load from, it drops the data from that save right into the list, acting just as though you'd imported from a file. None of this requires any new blocks or edits to a Scratch project itself!

That's one of the big principles here—I'm pretty sure suggestions for addons that add totally new behavior to the *player* itself will probably come under some scrutiny, especially a big idea like this. So, I've paid attention to make sure this doesn't do anything Scratch users can't *already* do with a project—it just makes the user interface and process a lot more streamlined. I think the biggest change this really makes is exposing controls based on the "import" and "export" options right to the player page; normally, you have to see inside to access those. So, project creators would still have to include a "if you don't see options to import and export the list, see inside and try again" type of note, if they want to make the save system work for people who don't use this addon. That should be included as a note in the addon manager. Otherwise, I think everything here is totally interoperable with the way people already add saves to their games!

**Extra notes**

* I'm thinking saves would be stored in the local browser storage, though there should probably be an option for exporting/importing some or all of your saves at once, so you can easily back 'em up or transfer them across devices.
* I figure saves will be grouped per project, to help keep things organized and simple for the user. That said, the option to import a save from a different project might be cool for multiple-part projects!
* All the stuff I'm talking about here *is* stuff I've seen in actual Scratch projects, though it's not *super* common. Another hope of this addon is that it might get people more aware of using lists as the backbone for save systems in general, prompting more folks to try out more creative uses of saves!

**Comment by Secret-chest** [comment]
I think it's rejected, it does not work for non-SA users.

**Comment by mxmou** [comment]
> I think it's rejected, it does not work for non-SA users.

Non-SA users could still see inside and import/export the list from there.

**Comment by mxmou** [comment]
A problem with this suggestion is that there is no way for the project to know when the user loads a previous save. It might be better if saving was done through lists and loading through ask prompts, similar to how many projects already do it.

**Comment by retronbv** [comment]
This might be able to work by saving all variables, then the next time the person views the project they can choose to load the saved cars and delete saved states

**Comment by towerofnix** [comment]
> I think it's rejected, it does not work for non-SA users.

@Secret-chest Like @mxmou said, players can still see inside to right-click the list and choose "import". This suggestion just streamlines the process around that.

> A problem with this suggestion is that there is no way for the project to know when the user loads a previous save. It might be better if saving was done through lists and loading through ask prompts, similar to how many projects already do it.

@mxmou Yes, that's a good point. This is something existing projects already have to deal with; like you said, a workaround is copying from lists and pasting into the "ask" prompt. If that's pretty popular, it's definitely reason to implement that way!

(Actually, I meant to include a note about that in the "extra notes" section (since it's sort of a follow-up to #1988). I forgot, oops. :P)

Maybe a floating button labelled "load save" (or just "load") above the check mark?

I do think it'd still be cool to keep the option to export/import entire lists, since that's a bit more versatile and is still possible for non-SA users to work with. A bit less practical, sure, but I think it could still lead to some creative new projects, and the note I recommended should keep creators from accidentally making the project unclear for users without the addon:

> So, project creators would still have to include a "if you don't see options to import and export the list, see inside and try again" type of note, if they want to make the save system work for people who don't use this addon. That should be included as a note in the addon manager.

> This might be able to work by saving all variables, then the next time the person views the project they can choose to load the saved cars and delete saved states

@retronbv This is another cool suggestion! I think a way to save the current project state would be neat. But there's no way to do the same thing in native Scratch, besides seeing inside and typing however many values into the appropriate "set variable" blocks. I tried to keep this suggestion pretty simple, working on just the existing capabilities of Scratch, so that it would automatically work with existing games, and wouldn't create anything new which is *only* possible for users of the addon.

Thanks all for the input!

**Comment by mxmou** [comment]
> I do think it'd still be cool to keep the option to export/import entire lists, since that's a bit more versatile and is still possible for non-SA users to work with. A bit less practical, sure, but I think it could still lead to some creative new projects, and the note I recommended should keep creators from accidentally making the project unclear for users without the addon.

Many existing projects, including [this one](https://scratch.mit.edu/projects/369597049/) I made, have separate UIs for saving (which show a list from where the user can copy the save code) and for loading (with an ask prompt for pasting the code). If lists have a load button, people might try to load saved data into the save UI and get confused because it probably won't work (most projects don't test if the list was changed when the user exits the save UI).

**Comment by towerofnix** [comment]
> Many existing projects, including [this one](https://scratch.mit.edu/projects/369597049/) I made, have separate UIs for saving (which show a list from where the user can copy the save code) and for loading (with an ask prompt for pasting the code). If lists have a load button, people might try to load saved data into the save UI and get confused because it probably won't work (most projects don't test if the list was changed when the user exits the save UI).

Oh, thanks! That does make sense as a confusing case for the user. Probably better to just let users do list-based import manually, then.

**Comment by towerofnix** [comment]
Pulling from discussion in #2013, here's another reason for restricting list-based import to the usual manual way (i.e. no new behavior there): cheating in multiplayer games! Being unexpectedly able to import into lists without entering the editor might be exploited in games that run on cloud data. (There are other ways to cheat cloud data, of course, but we don't need to make it any easier.)

**Comment by retronbv** [comment]
✂️
> > This might be able to work by saving all variables, then the next time the person views the project they can choose to load the saved cars and delete saved states
> 
> @retronbv This is another cool suggestion! I think a way to save the current project state would be neat. But there's no way to do the same thing in native Scratch, besides seeing inside and typing however many values into the appropriate "set variable" blocks. I tried to keep this suggestion pretty simple, working on just the existing capabilities of Scratch, so that it would automatically work with existing games, and wouldn't create anything new which is _only_ possible for users of the addon.

I mean saving the state in SA memory, not using scratch

**Comment by FunctionalMetatable** [comment]
This should follow a set of guidelines, e.g at line 2 of the list it says --sa-allow or something so the addon knows that this project supports said addon

**Comment by towerofnix** [comment]
@FunctionalMetatable Yes, I thought it might be valuable to only operate on lists which include (for example) the word "save" in their title. But I don't think it's strictly necessary; as long as the UI isn't too distracting, this doesn't (or no longer) gives access to anything that couldn't be done before, since per https://github.com/ScratchAddons/ScratchAddons/issues/1990#issuecomment-810638937 and https://github.com/ScratchAddons/ScratchAddons/issues/1990#issuecomment-809473254, you'd only be able to load into "ask and wait" prompts (which you can already type whatever you like into).

**Comment by lisa-wolfgang** [comment]
We would also want to implement it in a way that wouldn't attract negative ST attention -- if we're not careful, it could become common for projects to say "install the extension to save your progress" or something like that.

**Comment by Explosion-Scratch** [comment]
So would this basically save and load game state? It'd also have to do not just vars, but sound positions, sprite positions, costumes, etc. I could also see problems in this when the project is updated and saved data is attempted to be applied.

**Comment by mxmou** [comment]
> So would this basically save and load game state? It'd also have to do not just vars, but sound positions, sprite positions, costumes, etc.

It's just for saving data that can already be copied and pasted (list items), so the creator of the project can decide what information to encode in it.
